tests/network/services.go

```diff
@@ -0,0 +1,159 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package network
+
+import (
+	"fmt"
+	"strconv"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	batchv1 "k8s.io/api/batch/v1"
+	v12 "k8s.io/api/core/v1"
+	v13 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	v1 "kubevirt.io/client-go/api/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/kubevirt/tests"
+	cd "kubevirt.io/kubevirt/tests/containerdisk"
+)
+
+var _ = SIGDescribe("Services", func() {
+	var virtClient kubecli.KubevirtClient
+
+	runTCPClientExpectingHelloWorldFromServer := func(host, port, namespace string) *batchv1.Job {
+		job := tests.NewHelloWorldJob(host, port)
+		job, err := virtClient.BatchV1().Jobs(namespace).Create(job)
+		ExpectWithOffset(1, err).ToNot(HaveOccurred())
+		return job
+	}
+
+	BeforeEach(func() {
+		var err error
+		virtClient, err = kubecli.GetKubevirtClient()
+		Expect(err).NotTo(HaveOccurred(), "Should successfully initialize an API client")
+	})
+
+	Context("bridge interface binding", func() {
+		var inboundVMI *v1.VirtualMachineInstance
+		var serviceName string
+
+		const (
+			selectorLabelKey = "expose"
+			selectorLabelValue = "me"
+			servicePort = 1500
+		)
+
+		BeforeEach(func() {
+			// inboundVMI expects implicitly to be added to the pod network
+			inboundVMI = tests.NewRandomVMIWithEphemeralDiskAndUserdata(cd.ContainerDiskFor(cd.ContainerDiskCirros), "#!/bin/bash\necho 'hello'\n")
+			inboundVMI.Labels = map[string]string{selectorLabelKey: selectorLabelValue}
+			inboundVMI.Spec.Subdomain = "myvmi"
+			inboundVMI.Spec.Hostname = "my-subdomain"
+			// Remove the masquerade interface to use the default bridge one
+			inboundVMI.Spec.Domain.Devices.Interfaces = nil
+			inboundVMI.Spec.Networks = nil
+
+			_, err := virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(inboundVMI)
+			Expect(err).ToNot(HaveOccurred())
+
+			inboundVMI = tests.WaitUntilVMIReady(inboundVMI, tests.LoggedInCirrosExpecter)
+
+			tests.StartTCPServer(inboundVMI, servicePort)
+		})
+
+		Context("with a service matching the vmi exposed", func() {
+			BeforeEach(func() {
+				serviceName = "myservice"
+
+				service := buildServiceSpec(serviceName, servicePort, servicePort, selectorLabelKey, selectorLabelValue)
+				_, err := virtClient.CoreV1().Services(inboundVMI.Namespace).Create(service)
+				Expect(err).ToNot(HaveOccurred())
+			})
+
+			AfterEach(func() {
+				Expect(virtClient.CoreV1().Services(inboundVMI.Namespace).Delete(serviceName, &v13.DeleteOptions{})).To(Succeed())
+			})
```

Comment:
Don't we need to wait for the service to be deleted?

Reply:
We didn't do that before.

Reply:
We care about plenty of things, but I think we sometimes get the priorities of the things we care about wrong. 

Nothing will happen if we don't wait here - we'll wait when the VMI is torn down, which happens afterwards. And it takes a lot longer to delete the VM than in takes to delete this poor service. Can't we just dilute the waiting times between operations ?

Reply:
I know there have been issues with VMI deletions which have not been waited for, causing a large amount of VMI/s to accumulate together with their resources.
I do not know if there is a problem with Services or not, but your point is valid... if we wait for the VMI to get deleted, we should be fine (although not bullet proof).

Reply:
Has this been treated?

Reply:
How / Why ? Your last comment said "but your point is valid... if we wait for the VMI to get deleted, we should be fine".

What is the action point here ?